### PR TITLE
[BEAM-4294]  Join operator translator

### PR DESCRIPTION
**Pull request by VaclavPlajt**
Translation of Euphoria  Join suite operators (Inner, Left, Right, Full) into beam transformations. Implementation is centered around `JoinTranslator` which defines all the translation steps. And implementations of `JoinFn` which do the actual join on co-gruped elements, inspiration was drawn from `org.apache.beam.sdk.extensions.joinlibrary.Join`.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand:
   - [x] What the pull request does
   - [x] Why it does it
   - [x] How it does it
   - [x] Why this approach
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Run `./gradlew build` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    boolean rightHasValues = rightSIdeIter.iterator().hasNext();
+
+    if (leftHasValues && rightHasValues) {
+      for (RightT rightValue : rightSIdeIter) {
+        for (LeftT leftValue : leftSideIter) {
+          joiner.apply(leftValue, rightValue, outCollector);
+          c.output(Pair.of(key, outCollector.get()));
+        }
+      }
+    } else if (leftHasValues && !rightHasValues) {
```
Comment (mareksimunek) [inline sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java:42]: `!rightHasValues` is always true, you can delete it

File: sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      for (LeftT leftValue : leftSideIter) {
+        joiner.apply(leftValue, null, outCollector);
+        c.output(Pair.of(key, outCollector.get()));
+      }
+    } else if (!leftHasValues && rightHasValues) {
```
Comment (mareksimunek) [inline sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java:47]: !leftHasValues is always true, you can delete it

File: sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  @Override
+  public void processElement(ProcessContext c) {
+
+    KV<K, CoGbkResult> element = c.element();
+    CoGbkResult value = element.getValue();
+    K key = element.getKey();
+
+    Iterable<LeftT> leftSideIter = value.getAll(leftTag);
+    Iterable<RightT> rightSIdeIter = value.getAll(rightTag);
```
Comment (mareksimunek) [inline sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java:28]: variable typo. lowercase `i`

**Review by mareksimunek — COMMENTED** [review]
the beginning of `processElement` for specific Joins is the same. I could be moved to `JoinFn` which is their parent

File: sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java (hunk 1 above)
Comment (VaclavPlajt) [reply sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java:42]: Fixed in https://github.com/seznam/beam/commit/013b0783308fcd03d269b9fa1bf6c26e251d6169 This pull request will be closed.

File: sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java (hunk 2 above)
Comment (VaclavPlajt) [reply sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java:47]: Fixed in https://github.com/seznam/beam/commit/013b0783308fcd03d269b9fa1bf6c26e251d6169 This pull request will be closed.

File: sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java (hunk 3 above)
Comment (VaclavPlajt) [reply sdks/java/extensions/euphoria/euphoria-beam/src/main/java/cz/seznam/euphoria/beam/join/FullJoinFn.java:28]: Fixed in https://github.com/seznam/beam/commit/013b0783308fcd03d269b9fa1bf6c26e251d6169 This pull request will be closed.

**Comment by VaclavPlajt** [comment]
This pull request is replaced by #https://github.com/seznam/beam/pull/1

**Pull request closed without merging**